lib/drivers/bayerContourNext.js

```diff
@@ -0,0 +1,510 @@
+ * == BSD2 LICENSE ==
+ */
+
+/**
+ * IOET's code start here
+ * packet frame format <STX> FN text <ETX> C1 C2 <CR> <LF>
+ * */
+var _ = require('lodash');
+var async = require('async');
+var sundial = require('sundial');
+var crcCalculator = require('../crc.js');
+var struct = require('../struct.js')();
+var annotate = require('../eventAnnotations');
+
+var isBrowser = typeof window !== 'undefined';
+var debug = isBrowser ? require('../bows')('BCNextDriver') : debug;
+
+
+module.exports = function (config) {
+  var cfg = _.clone(config);
+  var hidDevice = config.deviceComms;
+
+  var STX = 0x02;
+  var ETB = 0x17;
+  var ETX = 0x03;
+  var EOT = 0x04;
+  var HID_PACKET_SIZE = 60;
+
+  /*
+   * TODO This info should be collected in the initial phase cause we can't assert if
+   *      it is the same for all contour devices
+   * */
+  var REPORT_BYTES = {
+    reportID: 0x41, // A
+    checksum: 0x00,
+    hostID  : 0x42, // B
+    deviceID: 0x43  // C
+  };
+
+  /* end */
+
+  var ASCII_CONTROL = {
+    ACK : 0x06,
+    NAK : 0x15,
+    ENQ : 0x05
+  };
+
+  var probe = function(cb){
+    debug('attempting probe of Bayer Contour Next');
+  };
+
+  var bcnPacketHandler = function (buffer) {
+    // we need to get the REPORT_BYTES here for the first communication
+    // time if they are not set
+    var discardCount = 3;
+    if (discardCount) {
+      buffer.discard(discardCount);
+    }
+
+    if (buffer.len() < 6) { // all complete packets must be at least this long
+      return false;       // not enough there yet
+    }
+
+    // there's enough there to try, anyway
+    var packet = extractPacket(buffer.bytes());
+    if (packet.packet_len !== 0) {
+      // remove the now-processed packet
+      buffer.discard(HID_PACKET_SIZE + 1);  // 1 is for packet_length
+    }
+
+    if (packet.valid) {
+      return packet;
+    } else {
+      return null;
+    }
+  };
+
+  var buildPacket = function (command, cmdlength) {
+    var datalen = cmdlength + 4; // We need just 5 bytes for now
+    var buf = new ArrayBuffer(datalen);
+    var bytes = new Uint8Array(buf);
+    var ctr = struct.pack(bytes, 0, 'bbb', REPORT_BYTES.reportID,
+                          REPORT_BYTES.hostID, REPORT_BYTES.deviceID,
+                          STX, command);
+    if (cmdlength) {
+      ctr += struct.pack(bytes, 3, 'bb', cmdlength, command);
+    }
+    return buf;
+  };
+
+  var buildReadCmd = function(recnum) {
+    return {
+        packet: buildPacket(ASCII_CONTROL.ACK, 1),
+        parser: function(result){
+          var tostr = _.map(result.bytes,
+                            function(e){
+                              return String.fromCharCode(e);
+                            }).join('');
+          result.payload = tostr;
+          return tostr;
+        }
+    };
+  };
+
+  var buildAckPacket = function() {
+    return buildPacket(ASCII_CONTROL.ACK, 1);
+  };
+
+  var buildHeaderPacket = function() {
+    var cmd = 0x3a;  // just anything 'x'
```

Comment:
The documentation states we should send anything to start the conversation, but at the end we are only using ACKs, I think we can remove this function safely.